src/checkpoint/dispatch/reconstructor.h

```diff
@@ -58,52 +59,172 @@ template <typename T>
   #if HAS_DETECTION_COMPONENT
     template <typename U>
     using isReconstructibleType =
-      typename std::enable_if<SerializableTraits<U>::is_reconstructible, T>::type;
+    typename std::enable_if<
+      SerializableTraits<U>::is_intrusive_reconstructible, T
+    >::type;
 
     template <typename U>
     using isNonIntReconstructibleType =
-      typename std::enable_if<
-        SerializableTraits<U>::is_nonintrusive_reconstructible, T
-      >::type;
+    typename std::enable_if<
+      SerializableTraits<U>::is_nonintrusive_reconstructible, T
+    >::type;
+
+    template <typename U>
+    using isNotReconstructibleType =
+    typename std::enable_if<not SerializableTraits<U>::is_reconstructible, T>::type;
+
+    template <typename U>
+    using isTaggedConstructibleType =
+    typename std::enable_if<SerializableTraits<U>::is_tagged_constructible, T>::type;
+
+    template <typename U>
+    using isNotTaggedConstructibleType =
+    typename std::enable_if<not SerializableTraits<U>::is_tagged_constructible, T>::type;
+
+    template <typename U>
+    using isConstructible =
+    typename std::enable_if<SerializableTraits<U>::is_constructible, T>::type;
+
+    template <typename U>
+    using isNotConstructible =
+    typename std::enable_if<not SerializableTraits<U>::is_constructible, T>::type;
   #else
     template <typename U>
-    using isNonIntReconstructibleType = isNotDefaultConsType<U>;
+    using isConstructible = isDefaultConsType<U>;
+
+    template <typename U>
+    using isNotConstructible = isNotDefaultConsType<U>;
   #endif
 
+  // Default-construct as lowest priority in reconstruction preference
   template <typename U = T>
-  static T* construct(void* buf, isDefaultConsType<U>* = nullptr) {
+  static T* constructDefault(void* buf, isDefaultConsType<U>* = nullptr) {
     debug_checkpoint("DeserializerDispatch: default constructor: buf=%p\n", buf);
     T* t_ptr = new (buf) T{};
     return t_ptr;
   }
 
+  // Fail, no valid option to constructing T
+  template <typename U = T>
+  static T* constructDefault(void* buf, isNotDefaultConsType<U>* = nullptr) {
+    #if HAS_DETECTION_COMPONENT
+    static_assert(
+      SerializableTraits<U>::is_tagged_constructible or
+      SerializableTraits<U>::is_reconstructible or
+      std::is_default_constructible<U>::value,
+      "Either a default constructor, reconstruct() function, or tagged "
+      "constructor are required for de-serialization"
+    );
+    #else
+    static_assert(
+      std::is_default_constructible<U>::value,
+      "A default constructor is required for de-serialization. To enable "
+      "reconstruct or tagged constructors, you must compile with the "
+      "detection component."
+    );
+    #endif
+    return nullptr;
+  }
+
   #if HAS_DETECTION_COMPONENT
+
+  /*
+   * Try to reconstruct with the following precedence:
+   *
+   *   - Tagged constructor: T(SERIALIZE_CONSTRUCT_TAG{})
+   *   - Reconstruct: T::reconstruct(buf) or reconstruct(t, buf)
```

Comment:
I think `reconstruct` ought to come ahead of a tagged constructor, but that doesn't need to be resolved here and now. I think we're never going to see the two of them in conflict. Maybe we should just detect and error if they're ever both available?